Feature: Allow manually placing town buildings in scenario editor.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Very little manual control over where houses are placed in towns. This is mostly to be expected as the game is meant to manage towns itself, however some manual control in the Scenario Editor (and perhaps a Sandbox mode) might be useful.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

* Add a network command to manually place houses. (Not strictly necessary for Scenario Editor but very useful for a sandbox mode.)
* Allow Houses to be resolved without a Town and Tile.
* Add a Town House builder to the Scenario Editor. This is accessed from the Landscaping toolbar as there is no town toolbar.

Once placed these houses behave like any other and can be removed by players and towns.

Uses the unified picker system, so also supports used/saved favourites. As town building don't have class labels, town zones are use to imitate them.

Houses sometimes have up to 4 variations, selected psuedo randomly based on tile. The House Selection window does not provide a way to choose between these, as there is no manual control, so in order to represent them in the UI it will cycle through each variation every 2.5 seconds.

"Some highly advanced town planning"
![image](https://github.com/OpenTTD/OpenTTD/assets/639850/1bedfaa7-d996-43ca-869b-be04a7d5f560)

More of the same, with NewGRFs...
![image](https://github.com/OpenTTD/OpenTTD/assets/639850/9660ce1e-e76b-4e0c-8963-6292a0f9cf4b)

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

This is limited to Scenario Editor only. No sandbox mode as been implemented as the Sandbox Settings window is still using the old cheats system.

Once that is changed to use normal settings, not much effort is required to enable this.


<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -2665,6 +2665,31 @@ static bool CheckTownBuild2x2House(TileIndex *tile, Town *t, int maxz, bool nosl
 	return false;
 }
 
+static void BuildTownHouse(Town *t, TileIndex tile, const HouseSpec *hs, HouseID house, uint8_t random_bits)
```
Missing Doxygen, and maybe should have a different name from the old `BuildTownHouse()`... Maybe the other should be renamed `TrySpawnTownHouse()` or similar?

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -2784,36 +2809,68 @@ static bool BuildTownHouse(Town *t, TileIndex tile)
+	if (!hs->enabled) return CMD_ERROR;
 
-		return true;
+	if (TimerGameCalendar::year < hs->min_year || TimerGameCalendar::year > hs->max_year) return CMD_ERROR;
```
A lot of these checks are duplicated from BuildTownHouse(). Could they be refactored into a shared `bool CheckTownHouseBuild()` or something?

PeterN replied:
They are very similar but with a few differences.

Some of the tests aren't performed because we let the user do more than towns.
TryBuildTownHouse just continues its loop, by CmdPlaceHouse needs to return error strings.

I did consider making TryBuildTownHouse issue a Command<CMD_PLACE_HOUSE>::Do. but that complicates CmdPlaceHouse more -- e.g TryBuildTownHouse only evaluates the slope and some other things once, not every attempt to build.

2TallTyler commented on src/newgrf_house.cpp:
```diff
@@ -105,11 +105,14 @@ void ResetHouses()
+	house_scope(*this, house_id, tile, town, not_yet_constructed, initial_random_bits, watched_cargo_triggers, view),
 	town_scope(*this, town, not_yet_constructed) // Don't access StorePSA if house is not yet constructed.
 {
+	/* Tile must be valid and a house tile, unless not yet construct in which case it may also be INVALID_TILE. */
```
```suggestion
	/* Tile must be valid and a house tile, unless not yet constructed in which case it may also be INVALID_TILE. */
```

2TallTyler commented on src/terraform_gui.cpp:
```diff
@@ -475,6 +476,8 @@ static constexpr NWidgetPart _nested_scen_edit_land_gen_widgets[] = {
 			NWidget(WWT_PUSHIMGBTN, COLOUR_GREY, WID_ETT_PLACE_OBJECT), SetMinimalSize(23, 22),
 										SetFill(0, 1), SetDataTip(SPR_IMG_TRANSMITTER, STR_SCENEDIT_TOOLBAR_PLACE_OBJECT),
+			NWidget(WWT_PUSHIMGBTN, COLOUR_DARK_GREEN, WID_ETT_PLACE_HOUSE), SetMinimalSize(22, 22),
+					SetFill(0, 1), SetDataTip(SPR_IMG_TOWN, STR_SCENEDIT_TOOLBAR_PLACE_OBJECT),
```
```suggestion
					SetFill(0, 1), SetDataTip(SPR_IMG_TOWN, STR_SCENEDIT_TOOLBAR_PLACE_HOUSE),
```

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -4061,3 +4062,98 @@ extern const TileTypeProcs _tile_type_town_procs = {
+		if (house_id >= NEW_HOUSE_OFFSET) {
+			/* Houses don't necessarily need new graphics. If they don't have a
+			* spritegroup associated with them, then the sprite for the substitute
+			* house id is drawn instead. */
```
```suggestion
			 * spritegroup associated with them, then the sprite for the substitute
			 * house id is drawn instead. */
```

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -4061,3 +4062,98 @@ extern const TileTypeProcs _tile_type_town_procs = {
+ * @param house_id House ID to draw.
+ * @param view The house's 'view'.
+ */
+void DrawNewHouseTileInGUI(int x, int y, const HouseSpec *spec, HouseID house_id, int view)
```
Should `DrawNewHouseTileInGUI()` and `DrawHouseInGUI()` be in `town_gui.cpp` instead of `town_cmd.cpp`, since they're related to the user interface and not commands?

2TallTyler commented on src/terraform_gui.cpp:
```diff
@@ -475,6 +476,8 @@ static constexpr NWidgetPart _nested_scen_edit_land_gen_widgets[] = {
 			EndContainer(),
 			NWidget(WWT_PUSHIMGBTN, COLOUR_GREY, WID_ETT_PLACE_OBJECT), SetMinimalSize(23, 22),
 										SetFill(0, 1), SetDataTip(SPR_IMG_TRANSMITTER, STR_SCENEDIT_TOOLBAR_PLACE_OBJECT),
+			NWidget(WWT_PUSHIMGBTN, COLOUR_DARK_GREEN, WID_ETT_PLACE_HOUSE), SetMinimalSize(22, 22),
```
```suggestion
			NWidget(WWT_PUSHIMGBTN, COLOUR_GREY, WID_ETT_PLACE_HOUSE), SetMinimalSize(22, 22),
```

2TallTyler reviewed (commented):
This is a great feature. A few thoughts:

* Adding this to Scenario Editor first and then adding it to normal gameplay as a later PR makes sense.
* Adding the button to the object toolbar works, but another option (which might make more sense) is adding it to the Town button dropdown. In SE there is no dropdown since there's only one option, but in-game you long-click the Town button to open a dropdown containing `Town directory` and `Found town`. This is where I would expect to find `Build houses`.

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -2784,36 +2817,68 @@ static bool TryBuildTownHouse(Town *t, TileIndex tile)
+
+	TileIndex origin_tile = tile;
+	if (hs->building_flags & TILE_SIZE_2x2) {
+		if (!CheckTownBuild2x2House(&tile, t, maxz, noslope)) return_cmd_error(STR_ERROR_CAN_T_CLEAR_THIS_AREA);
```
Should this (and the others nearby) return the error string `STR_ERROR_CAN_T_BUILD_HOUSE` instead?

PeterN replied:
No, it's a sub-error, although actually it was an incorrect string for a sub-error.

So I've changed this instead to actually call CMD_CLEAR for each tile to get a decent error message.

2TallTyler commented on src/town_gui.cpp:
```diff
@@ -1304,3 +1307,319 @@ void InitializeTownGui()
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_CLOSEBOX, COLOUR_DARK_GREEN),
+		NWidget(WWT_CAPTION, COLOUR_DARK_GREEN), SetDataTip(STR_HOUSE_PICKER_CAPTION, STR_TOOLTIP_WINDOW_TITLE_DRAG_THIS),
+		NWidget(WWT_DEFSIZEBOX, COLOUR_DARK_GREEN),
```
Other pickers have pin and shade buttons, but these have gone missing.

PeterN replied:
Only objects have the pin (people apparently like the object picker to stay open regardless of what toolbar is active).

Waypoint doesn't have the shadebox either, which is were I copied these widgets from, but I'll add it.

2TallTyler replied:
Ah, didn't notice the pin difference. That makes sense since only one transport type toolbar is visible at once -- if you have the road toolbar open and click on airports, that toolbar opens and the road toolbar is closed.

That said, houses don't have a transport type. Given that the house placer will likely be used for decorating cities similar to objects, I think it's appropriate for it to be pinnable. I can't speak to others' play styles, but I would personally like to pin it. 🙂

2TallTyler reviewed (commented):
I'd still prefer to see this accessed from a dropdown in the town button of the main toolbar, but if you feel strongly otherwise I'm open to having my mind changed. 😄

PeterN committed:
Add: Command to build an individual house on a specific tile.

PeterN committed:
Codechange: Rename BuildTownHouse to TryBuildTownHouse.

PeterN committed:
Codechange: Set underlying type for HouseSpec enums.

PeterN committed:
Change: Allow sprite resolver to evaluate non-existent houses.

This could previous be done but only with a town and tile in mind, but for drawing in the UI, neither of those exist yet.

PeterN committed:
Feature: Allow manually placing town buildings in scenario editor.

House picker is accessed from the Landscaping toolbar as there is no town toolbar.

Once placed these houses behave like any other and can be removed by players and towns.

Uses the unified picker system, so also supports used/saved favourites. As town building don't have class labels, town zones are use to imitate them.

2TallTyler commented on src/town_cmd.cpp:
```diff
@@ -2665,14 +2665,47 @@ static bool CheckTownBuild2x2House(TileIndex *tile, Town *t, int maxz, bool nosl
+ * @param tile The tile to try building on.
+ * @param hs The @a HouseSpec of the house.
+ * @param house The @a HouseID of the house.
+ * @param random_bits The random data to associated with the house.
```
```suggestion
 * @param random_bits The random data to be associated with the house.
```
